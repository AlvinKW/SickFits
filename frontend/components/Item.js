import React, { Component } from 'react';
import styled from 'styled-components';
import Link from 'next/link';
import PropTypes from 'prop-types';

import formatMoney from '../lib/formatMoney';

import DeleteItem from './DeleteItem';
import AddToCart from './AddToCart';

const StyledItem = styled.div`
	position: relative;
	display: flex;
	flex-direction: column;
	border: 1px solid ${props => props.theme.offWhite};
	box-shadow: ${props => props.theme.bs};
	background: white;

	img {
		height: 400px;
		width: 100%;
		object-fit: cover;
	}

	p {
		flex-grow: 1;
		padding: 0 3rem;
		font-size: 1.5rem;
		font-weight: 300;
		line-height: 2;
	}

	.buttonList {
		display: grid;
		grid-template-columns: repeat(auto-fit, minmax(100px, 1fr));
		grid-gap: 1px;
		width: 100%;
		border-top: 1px solid ${props => props.theme.lightGrey};
		background: ${props => props.theme.lightGrey};

		& > * {
			padding: 1rem;
			border: 0;
			font-family: "radnika_next";
			font-size: 1rem;
			background: white;
			color: ${props => props.theme.black};
			cursor: pointer;

			&:disabled {
				color: ${props => props.theme.lightGrey};
			}
		}
	}
`;

const StyledTitle = styled.h3`
	margin: 0 1rem;
	margin-top: -3rem;
	text-align: center;
	text-shadow: 2px 2px 0 rgba(0, 0, 0, 0.1);
	transform: skew(-5deg) rotate(-1deg);

	a {
		display: inline;
		padding: 0 1rem;
		font-size: 4rem;
		text-align: center;
		line-height: 1.3;
		background: ${props => props.theme.red};
		color: white;
	}
`;

const StyledPriceTag = styled.span`
	position: absolute;
	display: inline-block;
	top: -3px;
	right: -3px;
	padding: 5px;
	font-size: 3rem;
	font-weight: 600;
	line-height: 1;
	background: ${props => props.theme.red};
	color: white;
	transform: rotate(3deg);
`;

class Item extends Component {
	render() {
		const { item, permissions } = this.props;
		return (
			<StyledItem>
				{item.image && <img src={item.image} alt={item.title} />}
				<StyledTitle>
					<Link href={{ pathname: '/item', query: { id: item.id } }}>
						<a>{item.title}</a>
					</Link>
				</StyledTitle>
				<StyledPriceTag>{formatMoney(item.price)}</StyledPriceTag>
				<p>{item.description}</p>

				<div className="buttonList">
					{permissions.includes('ADMIN') && (
						<Link href={{ pathname: '/update', query: { id: item.id } }}>
							<a>Edit This Item</a>
						</Link>
					)}
					{permissions.includes('USER') && (
						<AddToCart id={item.id} />
					)}
					{permissions.includes('ADMIN') && (
						<DeleteItem id={item.id}>Delete This Item</DeleteItem>
					)}
				</div>
			</StyledItem>
		);
	}
}

Item.propTypes = {
	item: PropTypes.object.isRequired,
};

export default Item;
