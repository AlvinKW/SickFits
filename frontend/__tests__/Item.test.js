import { shallow } from 'enzyme';
import Item from '../components/Item';

const mockItem = {
	id: 'ABC123',
	title: 'A Cool Item',
	description: 'This item is really cool!',
	image: 'dog.jpg',
	largeImage: 'largeDog.jpg',
	price: 5000,
};

describe('<Item />', () => {
	it('renders the price-tag and title', () => {
		const wrapper = shallow(<Item item={mockItem} />);

		const PriceTag = wrapper.find('PriceTag');
		expect(PriceTag.children().text()).toBe('$50');

		expect(wrapper.find('Title a').text()).toBe(mockItem.title);
	});

	it('renders the image', () => {
		const wrapper = shallow(<Item item={mockItem} />);

		const img = wrapper.find('img');
		expect(img.props().src).toBe(mockItem.image);
		expect(img.props().alt).toBe(mockItem.title);
	});

	it('renders the buttons', () => {
		const wrapper = shallow(<Item item={mockItem} />);

		const buttonList = wrapper.find('.buttonList');
		expect(buttonList.children()).toHaveLength(3);

		expect(buttonList.find('Link')).toHaveLength(1);
		expect(buttonList.find('AddToCart')).toHaveLength(1);
		expect(buttonList.find('DeleteItem')).toHaveLength(1);
	});
});
