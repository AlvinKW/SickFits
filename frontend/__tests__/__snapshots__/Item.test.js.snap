// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Item /> renders and matches the snapshot 1`] = `
<Item__StyledItem>
  <img
    alt="A Cool Item"
    src="dog.jpg"
  />
  <Item__StyledTitle>
    <Link
      href={
        Object {
          "pathname": "/item",
          "query": Object {
            "id": "ABC123",
          },
        }
      }
    >
      <a>
        A Cool Item
      </a>
    </Link>
  </Item__StyledTitle>
  <Item__StyledPriceTag>
    $50
  </Item__StyledPriceTag>
  <p>
    This item is really cool!
  </p>
  <div
    className="buttonList"
  >
    <Link
      href={
        Object {
          "pathname": "/update",
          "query": Object {
            "id": "ABC123",
          },
        }
      }
    >
      <a>
        Edit This Item
      </a>
    </Link>
    <AddToCart
      id="ABC123"
    />
    <DeleteItem
      id="ABC123"
    >
      Delete This Item
    </DeleteItem>
  </div>
</Item__StyledItem>
`;
